### PR TITLE
Fix #3135 Shadow rendering darkening most buildings by rendering 'on top'.

### LINKED ISSUE #2397: 0.47.13 - odd rendering issue - "illuminated rocks"
0.47.13 standalone
Windows 10
Java 11.0.8

Nothing particularly major, just an odd rendering glitch, as some Rough hexes were in the shadow cast by high ground, and were fully in shadow, while in some other hexes that should have been equally affected, the base hex was shadowed, but the Rough 'boulders' were not in shadow. Did they have their own mood lighting?

Can be seen in attached screenshot under the north-east end of the bridge, while slightly to the north, there are some fully-shadowed Rough hexes. Also visible next to the narrowing of the gorge to the south.

### LINKED ISSUE #2675: 47.17 - Rendering Bug - 'Thin Snow' Terrain
OS: Windows 10
Java: AdoptOpenJDK 11.0.8

MegaMek 47.17 Standalone Install

(Issue also present in earlier versions.)

Odd quirk discovered with rendering hexes with 'Thin Snow' terrain modifier - shadows from woods, buildings, bridges, hills, etc are not rendering on 'Thin Snow', and neither are the 'cliff bottom graphics', or the lovely new 'incline edges'.

This suggests that 'Thin Snow' renders 'above' other graphics. (It also appears that 'Thin Snow' also obscures 'Rough' terrain, which may or may not be intentional.)

'Deep Snow' does not appear to have this issue, which is ... odd. :-)

See attached annotated screenshot. You will notice that the bridge's shadow suddenly vanished when it reaches the 'Thin Snow' hex, while it was still displayed on the hex with the road, which had 'Deep Snow' (to represent snowplough drifts, because this is me).

Thin Snow Shadow Issue Annotated

### LINKED ISSUE #3135: 0.49.3 - Shadow rendering darkening most buildings by rendering 'on top'
MegaMek 0.49.3 (MekHQ version)
Windows 10
Java OpenJDK 11

It appears that shadows on 0.49.3 are being rendered 'on top of' some buildings, making them appear very dark compared to previous versions of the game. This appears somewhat inconsistent, but appears to occur most often with Medium or heavier buildings, and with buildings of Height 3 or more - but also affects Light buildings adjacent to Medium buildings or Height 3 buildings.

With shadows ...
image

Without shadows ...
image

### PR DESCRIPTION
**Pull request by HoneySkull**
This is a "quick fix" for issue #3135.  When drawing the hexes with composite terrain such as pavement and buildings, the building is rendered under the shadow mask.  Most city maps have building on pavement and look particularly dark as described in issue #3135. 

Fixes #3135, fixes #2675, fixes #2397

### CODE REVIEW

**Comment by HoneySkull** [comment]
Before fix:
darkBuildingBefore

After fix: 
darkBuildingAfter

There still exists oddities here an there with pavement and building - but it's my opinion overall this appears much better for building on pavement in 99% of cases.

**Comment by HammerGS** [comment]
Big improvement in my opinion. Someone will get to the code review soon.

File: megamek/src/megamek/client/ui/swing/boardview/BoardView.java [hunk 1]
```diff
@@ -1431,6 +1430,8 @@ private void updateShadowMap() {
 
         Graphics2D g = shadowMap.createGraphics();
 
+        //Compute shadow angle based on planentary conditions.
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/boardview/BoardView.java:1433]:
```suggestion
        // Compute shadow angle based on planentary conditions.
```

**Review by Windchild292 — COMMENTED** [review]
One spacing issue, otherwise this looks good

**Comment by HoneySkull** [comment]
I'll fix the space issue.   Also found this - cleaning up as well:
darkBuildingRoads
darkBuildingRoadsAfter

**Comment by HoneySkull** [comment]
Good to go.

**Comment by HoneySkull** [comment]
Fixed: #2675 and #2397 which are shadow rendering related.

File: megamek/src/megamek/client/ui/swing/boardview/BoardView.java [hunk 2]
```diff
@@ -2509,6 +2513,14 @@ private void drawHex(Coords c, Graphics boardGraph, boolean saveBoardImage) {
             }
         }
 
+        // Check for buildings and woods burried under their own shadows.
+        if (supers != null && supersUnderShadow
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/boardview/BoardView.java:2517]:
```suggestion
        if ((supers != null) && supersUnderShadow
```

**Review by Windchild292 — COMMENTED** [review]
One new bracket issue, otherwise this looks good.

**Review by Windchild292 — APPROVED** [review]

**Pull request merged**